pints/_abc/_abc_smc.py

```diff
@@ -0,0 +1,198 @@
+            for parameter inference and model selection in dynamical systems.
+            Journal of the Royal Society Interface, 6.31: 187-202, 2009.
+            https://doi.org/10.1098/rsif.2008.0172
+
+    .. [2] "Beaumont, Mark A., et al. Adaptive approximate Bayesian
+            computation. Biometrika, 96.4: 983-990, 2009."
+            https://doi.org/10.48550/arXiv.0805.2256
+
+    Parameters
+    ----------
+    nr_samples
+        The number of samples requested for intermediate distributions.
+    error_schedule
+        The schedule of error threshold distance for all distributions.
+    """
+
+    def __init__(self, log_prior, perturbation_kernel=None):
+        # Log prior
+        self._log_prior = log_prior
+
+        # Default value for error threshold schedule
+        self._e_schedule = [1]
+
+        # Default value for current threshold
+        self._threshold = 1
+
+        # Size of intermediate distributions
+        self._nr_samples = 100
+
+        # Set up for first iteration
+        self._samples = [[]]
+        self._accepted_count = 0
+        self._weights = []
+        self._xs = None
+        self._ready_for_tell = False
+        self._t = 0
+        dim = log_prior.n_parameters()
+
+        # Setting the perturbation kernel
+        if perturbation_kernel is None:
+            self._perturbation_kernel = pints.MultivariateGaussianLogPrior(
+                np.zeros(dim),
+                0.001 * np.identity(dim))
+        elif isinstance(perturbation_kernel, pints.LogPrior):
+            self._perturbation_kernel = perturbation_kernel
+        else:
+            raise ValueError('Provided perturbation kernel must be an instance'
+                  ' of pints.LogPrior')
+
+    def name(self):
+        """ See :meth:`pints.ABCSampler.name()`. """
+        return 'ABC-SMC'
+
+    def ask(self, n_samples):
+        """ See :meth:`ABCSampler.ask()`. """
+        if self._ready_for_tell:
+            raise RuntimeError('ask called before tell.')
+        if self._t == 0:
+            self._xs = self._log_prior.sample(n_samples)
+        else:
+            self._xs = []
+
+            while len(self._xs) < n_samples:
+                theta_s_s = None  # to appease the linter
+                while (theta_s_s is None or
+                       self._log_prior(theta_s_s) == -np.inf):
+                    indices = np.random.choice(
+                        range(len(self._samples[self._t - 1])),
+                        p=self._weights[self._t - 1])
+                    theta_s = self._samples[self._t - 1][indices]
+                    # perturb using K_t
+                    theta_s_s = np.add(theta_s,
+                                       self._perturbation_kernel.sample(1)[0])
+                    # check if theta_s_s is possible under the prior
+                    # sample again if not
+                self._xs.append(theta_s_s)
+        self._ready_for_tell = True
+        return self._xs
```

Comment:
This allows users of the abc method to modify the values in self._xs, potentially corrupting everything. Return a numpy array instead and set it to read only (see e.g. the mcmc samplers or optimisers)